### PR TITLE
[Concurrency] Allow 'nonisolated' to be applied to mutable storage of 'Sendable' type on a globally-isolated value type.

### PR DESCRIPTION
**Pull request by simanerush**
rdar://130992526

Under SE-0434, `nonisolated` can be applied to mutable stored properties, as long as the property is on a globally-isolated value type and is of a `Sendable` type. 

Additionally, cross-module access should be allowed if such variable is marked `nonisolated`.

### CODE REVIEW

**Comment by simanerush** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckAttr.cpp [hunk 1]
```diff
@@ -7010,18 +7010,32 @@ void AttributeChecker::visitNonisolatedAttr(NonisolatedAttr *attr) {
...
-      if (var->supportsMutation() && !attr->isUnsafe()) {
-        diagnoseAndRemoveAttr(attr, diag::nonisolated_mutable_storage)
-            .fixItInsertAfter(attr->getRange().End, "(unsafe)");
-        var->diagnose(diag::nonisolated_mutable_storage_note, var);
-        return;
+      {
+        // 'nonisolated' can not be applied to mutable stored properties unless
+        // qualified as 'unsafe', or is of a Sendable type on a
+        // globally-isolated value type.
+        bool canBeNonisolated = false;
```
Comment (simanerush) [inline lib/Sema/TypeCheckAttr.cpp:7019]: This can probably have a better name, for now I just put it in a scope.

File: test/Concurrency/nonisolated_access.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+// Build the client
+// RUN: %target-swift-frontend -emit-module %t/src/Client.swift \
+// RUN: -module-name Client -I %t -swift-version 6 \
+// RUN: -emit-module-path %t/Client.swiftmodule
+
+// REQUIRES: concurrency
+
+//--- A.swift
+@preconcurrency @MainActor
```
Comment (xedin) [inline test/Concurrency/nonisolated_access.swift:17]: Do you actually need `@preconcurrency` here?

**Review by xedin — APPROVED** [review]

**Comment by simanerush** [comment]
@swift-ci please smoke test

File: test/Concurrency/actor_isolation.swift [hunk 3]
```diff
@@ -138,8 +138,7 @@ struct InferredFromContext {
     get { [] }
   }
 
-  nonisolated var status: Bool = true // expected-error {{'nonisolated' cannot be applied to mutable stored properties}}{{3-15=}}{{3-15=}}{{14-14=(unsafe)}}
```
Comment (AnthonyLatsis) [inline test/Concurrency/actor_isolation.swift:141]: Side note: This error message needs some love now that the rules have changed.

Comment (simanerush) [reply test/Concurrency/actor_isolation.swift:141]: I agree, but it's not obvious to me in what way should we change it. This case of allowing `nonisolated` on a mutable storage is too narrow, so updating the general case to include all conditions under which the attribute can be allowed seems too convoluted. I think I can see a possibility of updating this error to something like "'nonisolated' cannot be applied to this mutable stored property", and if it's declared on a globally-isolated value type, emit fix-its like "consider making the property type Sendable". Similarly, in the case where the property is of Sendable type and is on a value type, emit a fix-it saying "consider isolating the value type to the main actor." 
Another possibility would be splitting this into 2 diagnostics, one for value type and another for reference types. In that case, we _could_ add all the conditions in which this rule does not apply for the value type case.
In short, I'm just trying to think of a way to make this diagnostic make more sense from usability perspective (e.g. a programmer not familiar with subtleties in concurrency rules should get helpful fix-its)

File: lib/Sema/TypeCheckAttr.cpp [hunk 4]
```diff
@@ -7010,18 +7010,32 @@ void AttributeChecker::visitNonisolatedAttr(NonisolatedAttr *attr) {
...
+        if (dc->isTypeContext()) {
+          if (auto nominal = dc->getSelfStructDecl()) {
+            if (!var->isStatic() && type->isSendableType() &&
```
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply lib/Sema/TypeCheckAttr.cpp:7022]: Ok, proposal only applies to instance properties.
